test/counter_server_example/src/counter_server_example.erl

```diff
@@ -0,0 +1,261 @@
+  {ok, P1} = counter_server_supervisor:start_link_no_vnet(),
+  unlink(P1),
+  spawn(fun valid_client_no_vnet/0),
+  spawn(fun invalid_client_no_vnet/0).
+
+%%--------------------------------------------------------------------
+
+invalid_client_no_vnet() ->
+  ?assertEqual(error, local_request(invalid)).
+
+%%%-------------------------------------------------------------------
+
+%% @doc
+%% Same as before, but now the first client checks a property that is
+%% not always satisfied: each reply will be from the same generation
+%% of server. As the server might be broken and restarted, this
+%% assertion will fail in some interleavings.
+invalid_same_gen_no_vnet_test() ->
+  {ok, P1} = counter_server_supervisor:start_link_no_vnet(),
+  unlink(P1),
+  spawn(fun valid_client_same_gen_no_vnet/0),
+  spawn(fun invalid_client_no_vnet/0).
+
+%%--------------------------------------------------------------------
+
+valid_client_same_gen_no_vnet() ->
+  Res1 = valid_local_request(),
+  case Res1 =:= error of
+    true -> ok;
+    false ->
+      Res2 = valid_local_request(),
+      check_results(Res1, Res2),
+      check_same_gen(Res1, Res2)
+  end.
+
+%%--------------------------------------------------------------------
+
+check_same_gen(error, _) -> ok;
+check_same_gen(_, error) -> ok;
+check_same_gen({ok, G1, _R1}, {ok, G2, _R2}) ->
+  ?assertEqual(G1, G2);
+check_same_gen(Res1, Res2) ->
+  ?assertEqual(ok, {unexpected, Res1, Res2}).
+
+%%%===================================================================
+
+%% @doc
+%% Simulating a scenario with one node as client, making a request.
+once_valid_test() ->
+  Nodes = [server, client],
+  {ok, S} = vnet:start_link(Nodes),
+  unlink(S),
+  %% Synchronous server setup
+  vnet:rpc(server, counter_server_example, setup_server, []),
+  %% Async client request
+  ClientRPC = [counter_server_example, once_valid_client, []],
+  spawned_vnet_rpc(client, ClientRPC).
+
+%%%-------------------------------------------------------------------
+
+setup_server() ->
+  {ok, P} = counter_server_supervisor:start_link(),
+  unlink(P).
+
+%%%-------------------------------------------------------------------
+
+%% @doc Adding a spawn wrapper serves two purposes: it makes the
+%% request asynchronous and ensures exceptions reach the 'top'
+%% undisturbed and can be picked e.g. by --assertions_only (otherwise,
+%% vnet:rpc might use a try/catch statement and modify them).
+spawned_vnet_rpc(VNode, RPC) ->
+  vnet:rpc(VNode, erlang, spawn, RPC).
+
+%%%-------------------------------------------------------------------
+
+once_valid_client() ->
+  valid_rpc_request().
+
+valid_rpc_request() ->
+  server_request(increment).
+
+server_request(Request) ->
+  try
+    %% %% The commented out variant leads to schedulings where
+    %% %% vnet_proxy.erl, l52 crashes cause the process is not
+    %% %% alive and the call returns `undefined'
+    %% counter_server:request({via, vnet, {server, ?SERVER}}, increment)
```

Comment:
This one is on me to fix.

Actually, I think it would be beneficial to show both `rpc`+local `gen_server:call` and remote `gen_server:call` scenarios work equally under our model network.